pkg/subnet/local.go

```diff
@@ -123,54 +122,171 @@ func (d *Deployer) doDeploy(chain string, chainGenesis string) error {
 	}
 	chainID := genesis.Config.ChainID
 
-	customVMs := map[string]string{
-		chain: chainGenesis,
-	}
-
 	runDir := binutils.GetLatestRunDir()
 	if runDir == "" {
 		runDir = d.app.GetRunDir()
 	}
 
-	opts := []client.OpOption{
-		client.WithPluginDir(pluginDir),
-		client.WithCustomVMs(customVMs),
-		client.WithGlobalNodeConfig("{\"log-level\":\"debug\", \"log-display-level\":\"debug\"}"),
-		client.WithRootDataDir(runDir),
+	ctx := binutils.GetAsyncContext()
+
+	// check for network and get VM info
+	needsStart := false
+	clusterInfo, err := d.WaitForHealthy(ctx, cli, d.healthCheckInterval)
+	if err != nil {
+		// TODO: use error type not string comparison
+		if strings.Contains(err.Error(), "not bootstrapped") {
+			needsStart = true
+		} else {
+			return fmt.Errorf("failed to query network health: %s", err)
+		}
 	}
 
-	vmID, err := utils.VMID(chain)
+	// find out already deployed VM IDs
+	deployedVMIDs := map[string]struct{}{}
+	if clusterInfo != nil {
+		for _, vmInfo := range clusterInfo.CustomVms {
+			deployedVMIDs[vmInfo.VmId] = struct{}{}
+		}
+	}
+
+	// get VM ID to deploy
+	toDeployVMID, err := utils.VMID(chain)
 	if err != nil {
 		return fmt.Errorf("failed to create VM ID from %s: %w", chain, err)
 	}
-	d.app.Log.Debug("this VM will get ID: %s", vmID.String())
+	d.app.Log.Debug("this VM will get ID: %s", toDeployVMID.String())
 
+	// install all needed plugins
+	toInstallVMIDs := map[string]struct{}{}
+	for vmID := range deployedVMIDs {
+		toInstallVMIDs[vmID] = struct{}{}
+	}
+	toInstallVMIDs[toDeployVMID.String()] = struct{}{}
 	binDir := filepath.Join(d.app.GetBaseDir(), constants.AvalancheCliBinDir)
-	if err := d.binaryDownloader.Download(vmID, pluginDir, binDir); err != nil {
+	if err := d.binaryDownloader.Download(toInstallVMIDs, pluginDir, binDir); err != nil {
 		return err
 	}
 
-	ctx := binutils.GetAsyncContext()
+	ux.Logger.PrintToUser("VM ready.")
+
+	// do start/restart as needed
+	if needsStart {
+		ux.Logger.PrintToUser("Starting network...")
+		loadSnapshotOpts := []client.OpOption{
+			client.WithPluginDir(pluginDir),
+			client.WithExecPath(avalancheGoBinPath),
+			client.WithRootDataDir(runDir),
+		}
+		loadSnapshotsInfo, err := cli.LoadSnapshot(
+			ctx,
+			constants.DefaultSnapshotName,
+			loadSnapshotOpts...,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to start network :%s", err)
+		}
+		d.app.Log.Debug(loadSnapshotsInfo.String())
+	} else {
+		// we need to restart the network if the VM was not deployed previously
+		needsRestart := false
+		_, ok := deployedVMIDs[toDeployVMID.String()]
+		if !ok {
+			needsRestart = true
+		}
+
+		if needsRestart {
+			// make snapshot of current state and reload it again
+			tmpSnapshotName := fmt.Sprintf("restart-tmp-%d", time.Now().Unix())
```

Comment:
Do we really need a tmp snapshot? Can't we just override the snapshot and restart? It will be done anyways when we stop this network afterwards?